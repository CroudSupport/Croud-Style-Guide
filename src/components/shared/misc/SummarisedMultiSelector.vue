<template>
    <div class="dropdown-wrapper">
        <semantic-form-dropdown
            ref="summarisedSelector"
            v-show="!readOnly"
            v-model="model"
            :options="options"
            :data-tooltip="summary"
            :placeholder="placeholder"
            :data-inverted="true"
            :auto-update="false"
            :autoUpdateOptions="false"
            :full_text_search="true"
            :title_field="fields.title"
            :value_field="fields.value"
            :inline="inline"
            :fluid="fluid"
            search multiple
            :settings="dropdownSettings"
            @dropdown-selected="dropdownSelected"
        />

        <div v-if="readOnly && selectedLength < 1" :data-inverted="true">
            {{ placeholder }}
        </div>

        <div v-else-if="readOnly && selectedLength < 2" :data-tooltip="selectedItems" :data-inverted="true">
            {{ selectedItems }}
        </div>

        <div v-else-if="readOnly && selectedLength > 1" :data-tooltip="selectedItems" :data-inverted="true">
             {{ selectedLength }} Selected
        </div>
    </div>
</template>

<script>
    /**
     * A custom multi-select dropdown useful for multi selection scenarios
     *
     * @example ./croud-summarised-mutli-selector.md
     */

    export default {

        name: 'croud-summarised-multi-selector',

        props: {
            /**
            * Pass an array of options to display in the dropdown
            */
            options: {
                type: Array,
                default() {
                    return []
                },
            },

            /**
            * Maps the dropdown to the title/value fields on your options array.
            */
            fields: {
                type: Object,
                default() {
                    return {
                        title: 'name',
                        value: 'id',
                    }
                },
            },

            /**
            * The current value of the dropdown, v-model alias
            */
            value: {
                required: true,
                type: [String, Array],
                default() {
                    return ''
                },
            },

            /**
            * An optional placeholder for the dropdown
            */
            placeholder: {
                type: String,
                default() {
                    return ''
                },
            },

            /**
            * Default text for the dropdown summary popup
            */
            defaultSummary: {
                type: String,
                default() {
                    return 'Select an item'
                },
            },

            /**
            * Which field in the options to use in the summary popup
            */
            summaryField: {
                type: String,
                default() {
                    return 'name'
                },
            },

            /**
            * Semantic ui dropdown configuration
            */
            dropdownSettings: {
                type: Object,
                default() {
                    return {
                        useLabels: false,
                        showOnFocus: true,
                        onHide: () => this.$emit('dropdown-hidden'),
                    }
                },
            },

            /**
            * If the component should be in read only mode
            */
            readOnly: {
                type: Boolean,
                default() {
                    return false
                },
            },

            /**
            * If the component should be inline
            */
            inline: {
                type: Boolean,
                default() {
                    return true
                },
            },

            /**
            * If the component should be fluid
            */
            fluid: {
                type: Boolean,
                default() {
                    return true
                },
            },
        },

        computed: {
            model: {
                get() {
                    return this.value
                },

                set(val) {
                    this.$emit('input', val)
                },
            },

            computedModel() {
                return Array.isArray(this.model) ? this.model : this.model.split(',')
            },

            selectedLength() {
                return this.computedModel.length
            },

            selectedItems() {
                return this.options.filter((option) => {
                    if (this.computedModel.toString().indexOf(option[this.fields.value].toString()) > -1) {
                        return option[this.fields.value]
                    } return null
                }).map(selected => selected[this.summaryField]).join(', ')
            },

            summary() {
                if (this.model.length < 1) return this.defaultSummary
                return this.selectedItems
            },
        },

        methods: {
            dropdownSelected(value) {
                this.$emit('dropdown-selected', value)
            },

            setLabel() {
                if (this.model.length < 1) return
                if (this.selectedLength < 2) {
                    $(this.$refs.summarisedSelector.$el).dropdown('set text', this.selectedItems)
                } else {
                    $(this.$refs.summarisedSelector.$el).dropdown('set text', `${this.selectedLength} Selected`)
                }
            },

            reset() {
                $(this.$refs.summarisedSelector.$el).dropdown('clear')
                this.$nextTick(() => {
                    $(this.$refs.summarisedSelector.$el).dropdown('refresh')
                    this.$nextTick(this.$forceUpdate())
                })
            },
        },

        watch: {
            model() {
                this.setLabel()
            },
        },

        mounted() {
            this.$nextTick(this.setLabel())
        },

    }
</script>

<style scoped lang="scss">
    @import '../../../resources/sass/variables/_all.scss';

    .dropdown-wrapper /deep/ .ui.dropdown.search.fluid.multiple.inline {
        padding: 0;

        .menu {
            overflow-x: visible;
            overflow-y: visible;
            max-height: initial;
        }

        span.sizer {
            display: none;
        }
    }

    [data-tooltip]:after {
        z-index: $croud-layer-2;
    }
</style>
