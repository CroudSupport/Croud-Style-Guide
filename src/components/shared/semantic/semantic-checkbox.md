### semantic-checkbox
For our checkboxes, we use the semantic-checkbox component from the [croud-vue-semantic](https://github.com/CroudSupport/vue-semantic) library. Below are a few examples of semantic-checkbox in action. Click [here](http://croudsupport.github.io/vue-semantic/#semantic-checkbox) for full component docs.

### Basic usage

    <semantic-checkbox v-model="checkbox" label="🍕 Pizza" />

### Styling
Pass through the **type** prop to change the style of checkbo

    <div class="ui form">
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Checkbox" />
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Slider" type="slider"/>
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Toggle" type="toggle" />
        </div>
    </div>

### Disabled

    <semantic-checkbox label="🍕 Pizza" type="toggle" :disabled="true"/>
