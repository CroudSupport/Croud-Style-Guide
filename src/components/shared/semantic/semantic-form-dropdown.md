### semantic-form-dropdown
For our form dropdowns, we use the semantic-form-dropdown component from the [croud-vue-semantic](https://github.com/croudtech/vue-semantic) library. Below are a few examples of semantic-form-dropdown in action. Click [here](http://croudtech.github.io/vue-semantic/#semantic-form-dropdown) for full component docs.

### Basic usage

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]"></semantic-form-dropdown>

### Styling
You can use the props to change the styling of the dropdown

    <div class="ui form">
        <div class="field">
            <label>Fluid</label>
            <semantic-form-dropdown fluid v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>

        <div class="three fields">
            <div class="field">
                <label>Search</label>
                <semantic-form-dropdown search v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Multiple</label>
                <semantic-form-dropdown multiple v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Disabled</label>
                <semantic-form-dropdown :disabled="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
        </div>

        <div class="inline field">
            <label>Inline</label>
            <semantic-form-dropdown :inline="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>
    </div>

### Using slots
The **default** slot allows you to add custom options to the top of the dropdown

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]">
        <div class="item" v-data="3">
            <b>Baz</b>
        </div>
    </semantic-form-dropdown>

### Handle Events
This component emits the **dropdown-selected** event when an item is selected.

    <semantic-form-dropdown :value="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" @dropdown-selected="$toasted.success('TaDa!!!')"></semantic-form-dropdown>
