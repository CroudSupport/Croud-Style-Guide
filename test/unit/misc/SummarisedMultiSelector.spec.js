import Vue from 'vue'
import VueSemantic from 'croud-vue-semantic'
import SummarisedMultiSelector from '../../../src/components/shared/misc/SummarisedMultiSelector'
import '../../../semantic/dist/semantic'

Vue.use(VueSemantic)

const Constructor = Vue.extend(SummarisedMultiSelector)

const vm = new Constructor({
    propsData: {
        readOnly: false,
        options: [
            { id: 1, name: 'PPC' },
            { id: 2, name: 'Paid Social' },
            { id: 3, name: 'SEO' },
        ],
    },
}).$mount()

describe('summarised multi selector', () => {
    describe('none selected', () => {
        it('should match the snapshot', () => {
            expect(vm.$el).toMatchSnapshot()
        })

        describe('read only', () => {
            beforeEach(() => {
                vm.readOnly = true
            })

            it('should match the snapshot', () => {
                expect(vm.$el).toMatchSnapshot()
            })
        })
    })

    describe('some selected', () => {
        describe('build summary', () => {
            describe('from array', () => {
                it('should handle single value', () => {
                    vm.value = [1]
                    vm.$nextTick(() => {
                        expect(vm.selectedItems).toBe('PPC')
                        expect(vm.displayText).toBe('PPC')
                    })
                })

                it('should handle mutliple values', () => {
                    vm.value = [1, 2]
                    vm.$nextTick(() => {
                        expect(vm.selectedItems).toBe('PPC, Paid Social')
                        expect(vm.displayText).toBe('2 Selected')
                    })
                })

                it('should match the snapshot', () => {
                    vm.value = [1, 2]
                    expect(vm.$el).toMatchSnapshot()
                })
            })

            describe('from string', () => {
                it('should handle single value', () => {
                    vm.value = '1'
                    vm.$nextTick(() => {
                        expect(vm.selectedItems).toBe('PPC')
                        expect(vm.displayText).toBe('PPC')
                    })
                })

                it('should handle mutliple values', () => {
                    vm.value = '1,2'
                    vm.$nextTick(() => {
                        expect(vm.selectedItems).toBe('PPC, Paid Social')
                        expect(vm.displayText).toBe('2 Selected')
                    })
                })

                it('should match the snapshot', () => {
                    vm.value = '1,2'
                    expect(vm.$el).toMatchSnapshot()
                })
            })

            describe('with different summary field', () => {
                it('should use the id field', () => {
                    vm.value = '1'
                    vm.fields = {
                        summary: 'id',
                    }
                    vm.$nextTick(() => {
                        expect(vm.selectedItems).toBe('1')
                        expect(vm.displayText).toBe('1')
                    })
                })
            })
        })

        describe('read only', () => {
            beforeEach(() => {
                vm.readOnly = true
                vm.value = '1,2'
                vm.fields = {}
            })

            it('should match the snapshot', () => {
                expect(vm.$el).toMatchSnapshot()
            })
        })
    })
})
