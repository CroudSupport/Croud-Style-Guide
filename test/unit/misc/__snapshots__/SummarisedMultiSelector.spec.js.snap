// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`summarised multi selector read only should match the snapshot 1`] = `
<div
  class="dropdown-wrapper"
>
  <div
    data-tooltip=""
    data-inverted="true"
  >
    
  </div>
</div>
`;

exports[`summarised multi selector should match the snapshot 1`] = `
<div
  class="dropdown-wrapper"
>
  <div
    class="ui dropdown search fluid multiple inline"
    data-tooltip="Select an item"
    data-inverted="true"
  >
    <input
      type="hidden"
      name=""
    />
    <input
      class="search"
      autocomplete="off"
      tabindex="0"
    />
    <span
      class="sizer"
    />
    <div
      class="default text"
    >
      
    </div>
    <i
      class="dropdown icon"
    />
    <div
      class="menu dropdown_menu"
      tabindex="-1"
    >
      <div
        data-value="1"
        data-text="PPC"
        class="item"
      >
        PPC
      </div>
      <div
        data-value="2"
        data-text="Paid Social"
        class="item"
      >
        Paid Social
      </div>
      <div
        data-value="3"
        data-text="SEO"
        class="item"
      >
        SEO
      </div>
    </div>
  </div>
</div>
`;
